File: Pick,fit(4F)/Back-end/PickFit/src/main/java/com/pickfit/pickfit/config/SecurityConfig.java
package com.pickfit.pickfit.config;

import com.pickfit.pickfit.oauth2.handler.OAuth2AuthenticationSuccessHandler;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpMethod;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.web.servlet.config.annotation.CorsRegistry;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

@Configuration
@EnableWebSecurity
public class SecurityConfig implements WebMvcConfigurer {

    private final OAuth2AuthenticationSuccessHandler successHandler;

    public SecurityConfig(OAuth2AuthenticationSuccessHandler successHandler) {
        this.successHandler = successHandler;
    }

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        http
                .cors() // CORS 설정 활성화
                .and()
                .csrf().disable() // CSRF 보호 비활성화 (FastAPI와의 통신 테스트를 위해 비활성화)
                .authorizeHttpRequests(authorize -> authorize
                        .requestMatchers("/api/**").permitAll() // `/api/**` 경로는 모두 인증 없이 접근 가능
                        .requestMatchers(HttpMethod.OPTIONS, "/**").permitAll() // OPTIONS 요청 허용 (CORS 사전 요청)
                        .anyRequest().permitAll() // 모든 요청 인증 없이 허용
                )
                .oauth2Login(oauth2 -> oauth2
                        .successHandler(successHandler) // OAuth2 로그인 성공 핸들러 설정
                );

        return http.build();
    }

    @Override
    public void addCorsMappings(CorsRegistry registry) {
        registry.addMapping("/**") // 모든 경로에 대해 CORS 허용
                .allowedOrigins("http://localhost:3000", "http://localhost:5000", "http://pickfit.store", "http://52.79.200.245:5688", "http://52.79.200.245") // React, FastAPI 허용
                .allowedMethods("GET", "POST", "PUT", "DELETE", "OPTIONS") // 허용할 HTTP 메서드
                .allowedHeaders("*") // 모든 헤더 허용
                .allowCredentials(true); // 쿠키 및 인증 정보 허용
    }
}

File: Bask-EAT((주)식회사)/Bask-EAT-Service-Backend/bin/main/com/bask_eat/business_backend/config/SecurityConfig.kt
package com.bask_eat.business_backend.config


import com.bask_eat.business_backend.security.CustomUserDetails
import com.bask_eat.business_backend.security.FirebaseTokenFilter
import com.bask_eat.business_backend.security.JwtAuthenticationFilter
import com.bask_eat.business_backend.security.JwtTokenProvider
import com.bask_eat.business_backend.service.UserService
import kotlinx.coroutines.runBlocking
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpMethod
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.AuthenticationProvider
import org.springframework.security.authentication.dao.DaoAuthenticationProvider
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.AuthenticationEntryPoint
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.access.AccessDeniedHandler
import org.springframework.security.web.authentication.AuthenticationFailureHandler
import org.springframework.security.web.authentication.AuthenticationSuccessHandler
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.security.web.context.HttpSessionSecurityContextRepository
import org.springframework.security.web.context.SecurityContextRepository
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import org.springframework.web.cors.UrlBasedCorsConfigurationSource

@Configuration
@EnableWebSecurity
@EnableMethodSecurity
class SecurityConfig(
  private val firebaseTokenFilter: FirebaseTokenFilter,
  private val jwtAuthenticationFilter: JwtAuthenticationFilter,
  private val jwtTokenProvider: JwtTokenProvider,
  private val userService: UserService
) {

  @Bean
  fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
  return http
      .securityMatcher("/api/**", "/fh/api/**")
      .csrf { it.disable() }
      .sessionManagement { it.sessionCreationPolicy(SessionCreationPolicy.STATELESS) }
      .headers { headers ->
        headers.frameOptions().deny()
      }
      .authorizeHttpRequests { auth ->
        auth
          .requestMatchers(HttpMethod.OPTIONS, "/**").permitAll()
          .requestMatchers(HttpMethod.POST, "/api/auth/**", "/fh/api/auth/**").permitAll()
          .requestMatchers("/api/health").permitAll()
          .anyRequest().authenticated()
      }
      .addFilterBefore(firebaseTokenFilter, UsernamePasswordAuthenticationFilter::class.java)
      .addFilterBefore(jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter::class.java)
      .oauth2Login { oauth2 ->
        oauth2
          .successHandler(oauth2AuthenticationSuccessHandler())
          .failureHandler(oauth2AuthenticationFailureHandler())
      }
      .exceptionHandling { exceptions ->
        exceptions
          .authenticationEntryPoint(customAuthenticationEntryPoint())
          .accessDeniedHandler(customAccessDeniedHandler())
      }
      .cors { it.configurationSource(corsConfigurationSource()) }
      .securityContext { spec ->
        spec.securityContextRepository(securityContextRepository())
      }
      .build()
  }

  @Bean
  fun oauth2AuthenticationSuccessHandler(): AuthenticationSuccessHandler {
    return AuthenticationSuccessHandler { _, response, _ ->
      response.sendRedirect("/api/auth/success")
    }
  }

  @Bean
  fun oauth2AuthenticationFailureHandler(): AuthenticationFailureHandler {
    return AuthenticationFailureHandler { _, response, _ ->
      response.sendRedirect("/api/auth/error")
    }
  }

  @Bean
  fun customAuthenticationEntryPoint(): AuthenticationEntryPoint {
    return AuthenticationEntryPoint { _, response, _ ->
      response.sendError(401, "Unauthorized")
    }
  }

  @Bean
  fun customAccessDeniedHandler(): AccessDeniedHandler {
    return AccessDeniedHandler { _, response, _ ->
      response.sendError(403, "Access Denied")
    }
  }

  @Bean
  fun passwordEncoder(): PasswordEncoder {
    return BCryptPasswordEncoder()
  }

  @Bean
  fun corsConfigurationSource(): CorsConfigurationSource {
    val configuration = CorsConfiguration()
    configuration.allowedOriginPatterns = listOf("*")
    configuration.allowedMethods = listOf("GET", "POST", "PUT", "DELETE", "OPTIONS")
    configuration.allowedHeaders = listOf("*")
    configuration.exposedHeaders = listOf("Authorization", "X-Access-Token")
    configuration.allowCredentials = true

    val source = UrlBasedCorsConfigurationSource()
    source.registerCorsConfiguration("/**", configuration)
    return source
  }

  @Bean
  fun authenticationManager(http: HttpSecurity): AuthenticationManager {
    // UserDetailsService와 PasswordEncoder를 사용하지 않을 경우,
    // 빈 AuthenticationProvider가 없으면 빈으로 반환되지 않음.
    return http
      .getSharedObject(AuthenticationManagerBuilder::class.java)
      .authenticationProvider(jwtAuthenticationProvider())
      .build()
  }

  @Bean
  fun jwtAuthenticationProvider(): AuthenticationProvider {
    val provider = DaoAuthenticationProvider()
    provider.setUserDetailsService { username ->
      val user = runBlocking { userService.findById(username) }
        ?: throw UsernameNotFoundException("User not found: $username") as Throwable
      CustomUserDetails(user)
    }
    provider.setPasswordEncoder(passwordEncoder())
    return provider
  }

  @Bean
  fun securityContextRepository(): SecurityContextRepository =
    HttpSessionSecurityContextRepository()


}
